### PR TITLE
Design of if else op

### PR DESCRIPTION
**Pull request by zchen0211**

### CODE REVIEW

File: doc/design/if_else_op.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
+In an if_op, only inputs with condition satisfied will be run. The op could have multiple inputs and multiple outputs.
+We should have the following design:
+
+```python
```
Comment (wangkuiyi) [inline doc/design/if_else_op.md]: Basing on https://github.com/PaddlePaddle/Paddle/pull/3827, Python bindings of operators and layers should return Vars. So I think that this usage could be

```python
x1 = Var()
x2 = Var()
y1 = Var()
y2 = Var()
cond = Var()

with paddle.block() as left:
  x1 = if_input(x1)
  x2 = if_input(x2)
  y1 = mul(x1, x2)
  y2 = add(x1, x2)

with paddle.block() as right:
  x1 = if_input(x1)
  x2 = if_input(x2)
  y1 = add(x1, x2)
  y2 = mul(x1, x2)

y1, y2 = layer.if(cond, left, right) # This line is equal to layer.if(cond, left, right)

File: doc/design/if_else_op.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+with if.true_block() as block:
+  x1 = if.input(x1)
+  x2 = if.input(x2)
+  y = pd.add(x1, x2)
+  y2 = pd.fc(x1) # contains (w,b)
+  if.output(y)
+  if.output(y2)
+  
+o1, o2 = if(cond)
```
Comment (wangkuiyi) [inline doc/design/if_else_op.md]: No branch specified?  I think this needs to be

```python
o1, o2 = layer.if(cond, block, output=[o1, o2])
```

File: doc/design/if_else_op.md (hunk 1 above)
Comment (Superjom) [reply doc/design/if_else_op.md]:
```python
import paddle as pd

x1 = Var()
x2 = Var()
y1 = Var()
y2 = Var()
cond = Var()

c1 = pd.IfElseOp(cond, inputs=[x1, x2], outputs=[y1, y2])
with c1.true_block():
    x1, x2 = c1.inputs
    y1, y2 = c1.outputs

    y1 = mul(x1, x2)
    y2 = add(x1, x2)

with c1.false_block():
    x1, x2 = c1.inputs
    y1, y2 = c1.outputs

    y1 = add(x1, x2)
    y2 = mul(x1, x2)
```

@wangkuiyi

Comment (wangkuiyi) [reply doc/design/if_else_op.md]: I believe @Superjom 's proposal is better than mine!

Comment (wangkuiyi) [reply doc/design/if_else_op.md]: [Above example](https://github.com/PaddlePaddle/Paddle/pull/3828#discussion_r136707032) breaks the convention that Python operator bindings must return Vars.  

However, [mine](https://github.com/PaddlePaddle/Paddle/pull/3828#discussion_r136706744) has a severe problem that the two branches would overwrite y1 and y2. A right solution should keep two copies of y1 and y2 for the left and the right branches respectively and then merge these two copies into the global variables y1 and y2.

A correct solution can be derived easily from @Superjom 's solution -- to rename `paddle.IfElseOp` into `paddle.CreateIfElseOpBuilder`.

Comment (wangkuiyi) [reply doc/design/if_else_op.md]:
```python
import paddle as pd

x1 = Var()
x2 = Var()
cond = Var()
c = pd.create_ifelseop_builder(inputs=[x1, x2], output_num=2)
with c.true_block() as b:
    x1, x2 = b.inputs()
    b.set_output(0, mul(x1, x2))
    b.set_output(1, add(x1, x2))

with c.false_block() as b:
    x1, x2 = c1.inputs()
    b.set_output(0, add(x1, x2))
    b.set_output(1, mul(x1, x2))

out1, out2 = c(cond)
```

File: doc/design/if_else_op.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+```
+
+In an if_op, only inputs with condition satisfied will be run.
+We should have the following design:
+```python
```
Comment (wangkuiyi) [inline doc/design/if_else_op.md]: IfOp should have only one branch.  An IfOp operator takes a `cond` variable whose value must be a vector of N boolean elements. Its return value has M (M<=N) instances, each corresponds to a true element in `cond`.

```python
import paddle as pd

x = var()
y = var()
cond = var()

b = pd.create_ifop_builder(inputs=[x], output_num=1)
with b.true_block():
    x = b.inputs(0)
    z = operator.add(x, y)
    b.set_output(0, operator.softmax(z))

out = b(cond)
```

If we want the output still has N instances, we can use IfElseOp with a default value, whose minibatch size must be N:

```python
import paddle as pd

x = var()
y = var()
cond = var()
default_value = var()
b = pd.create_ifelseop_builder(inputs=[x], output_num=1, default_value)
with b.true_block():
    x = b.inputs(0)
    z = operator.add(x, y)
    b.set_output(0, operator.softmax(z))

out = b(cond)
```

File: doc/design/if_else_op.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
+IfOp should have only one branch. An IfOp operator takes a `cond` variable whose value must be a vector of N boolean elements. Its return value has M (M<=N) instances, each corresponds to a true element in `cond`.
```
Comment (Superjom) [inline doc/design/if_else_op.md]: should we add a static IfElseOp just like TF, just run one branch.

This design describe a DynamicIfElseOp ? am I right?

Comment (wangkuiyi) [reply doc/design/if_else_op.md]: Let's make this IfElseOp. If you are going to add a static conditional branching structure later, we can name it StaticIfElseOp.

File: doc/design/if_else_op.md [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+```python
+import paddle as pd
+
+x = var()
+y = var()
+cond = var()
+
+b = pd.create_ifop_builder(inputs=[x], output_num=1)
```
Comment (wangkuiyi) [inline doc/design/if_else_op.md]: create_ifop_builder ==> create_ifop, as suggested by @Superjom

File: doc/design/if_else_op.md [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+If we want the output still has N instances, we can use IfElseOp with a default value, whose minibatch size must be N:
+
+```python
+import paddle as pd
+
+x = var()
+y = var()
+cond = var()
+default_value = var()
+b = pd.create_ifelseop_builder(inputs=[x], output_num=1)
```
Comment (wangkuiyi) [inline doc/design/if_else_op.md]: create_ifelseop_builder => create_ifelseop

**Pull request merged**